**Fill the Code blocks according to the comments provided as hints**

In [1]:
from sklearn.datasets import load_diabetes #We will use this dataset
#import libraries for numpy, Linear regression, r2_score, train_test_split
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [2]:
X,y = load_diabetes(return_X_y=True)  #loading dataset

In [3]:
#print shape of X and y and write in a text block what do you observe about both 
print(f"X shape={X.shape}")
print(f"y shape={y.shape}")

X shape=(442, 10)
y shape=(442,)


shape of x tells us there are 442 rows and 10 columns, meaning 442 observations and 10 features each

shape of y tells us that y is a 1D array with 442 elements, which means there are 442 target values corresponding to each of the 442 observations

In [4]:
#split the dataset into Train(80%) and Test(20%)
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2)

In [5]:
#fit X,y in Linear Regression Model of sklearn
reg = LinearRegression()
reg.fit(X_train,y_train)

LinearRegression()

In [6]:
#print coefficients and intercept (in @d gradient descent in was only m and b but as it is n dimensional it will have n coefficients and 1 intercept)
print(f"coefficients={reg.coef_}")
print("\n")
print(f"intercept={reg.intercept_}")

coefficients=[  -4.52701172 -261.22022943  571.11340786  315.87662254 -838.17128955
  551.32170702   95.16917117  125.56564734  802.37156385   23.13661528]


intercept=151.78007216354953


In [7]:
#predict on X_test and calculate r2_score
y_pred = reg.predict(X_test)
r2_score(y_test,y_pred)

0.4535641886503101

In [8]:
#print X_train shape
print(f"X_train shape={X_train.shape}")

X_train shape=(353, 10)


In [9]:
#Write GDregressor class


class GDRegressor:
    
    def __init__(self,learning_rate,epochs):
        
        self.coef_ = None
        self.intercept_ = None
        self.lr = learning_rate
        self.epochs = epochs
        
    def fit(self,X_train,y_train):
        # init your coefs
        self.intercept_ =  0         #initialise the intercept
        self.coef_ = np.zeros(X_train.shape[1])         #initialise the array of coefficients according to number of coefficients
        
        for i in range(self.epochs):
            # update all the coef and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_    #apply prediction formula i.e. (y_hat = coefficient*X + intercept) in terms of matrices coefficient and X   
            intercept_derivative = -2 * np.sum(y_train - y_hat)           #find derivative matrix                        
            self.intercept_ = self.intercept_ - (self.lr * ( intercept_derivative))   #update intercept using above intercept_derivative  
            
            coef_derivative = -2 *(np.dot(X_train.T,y_train-y_hat))   #derivative matrix for all coefficients in a single step        
            self.coef_ = self.coef_ - (self.lr * coef_derivative)      #update intercept using coef_derivative                              
        print(f"coefficient:{self.coef_}")
        print(f"intercept:{self.intercept_}")     #print the coefficients array and the intercept
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_  #new prediction in the form of (y = coefficients*X + intercept) using matrices

In [10]:
#call GDRegressor by keeping epochs and learning rate according to you and keep changing it for better results
gd = GDRegressor(0.002,1000)

In [11]:
#fit X_train, y_train into GDRegressor object using function of the class
gd.fit(X_train,y_train)

coefficient:[  -4.26774219 -240.4398934   564.73789627  307.72691106  -46.77139503
 -110.93745156 -203.07086815  113.7729202   448.19298564   68.55285239]
intercept:151.9553204712101


In [12]:
#predict on X_test using function of the class
y_pred = gd.predict(X_test)

In [13]:
#calculate r2_score to verify
r2_score(y_test,y_pred)

0.4635727964703802